File: ArraysHashing/valid_sudoku.java
import java.util.HashSet;

public class valid_sudoku {
    @SuppressWarnings("unchecked") // just to surpress a warning given on run time
    public static boolean isValidSudoku(char[][] board) {
        // we want to create an array of hashsets one for rows, one for columns, and one for the boxes in sudoku and we know each will hold 9 elements
        HashSet<Character>[] rows = new HashSet[9];
        HashSet<Character>[] columns = new HashSet[9];
        HashSet<Character>[] boxes = new HashSet[9];

        for (int i = 0; i < 9; i++) { // so we initialize the hashsets
            rows[i] = new HashSet<>();
            columns[i] = new HashSet<>();
            boxes[i] = new HashSet<>();
        }

        for (int i = 0; i < 9; i++) { // double for loop to iterate over each cell in the board
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') { // if the cell is empty just continue to next element
                    continue;
                }

                if (rows[i].contains(board[i][j])) { // if the current row already contains this element then its false
                    return false;
                }
                rows[i].add(board[i][j]); // add to the current row otherwise

                if (columns[j].contains(board[i][j])) { // if the current column already contains this element then its false
                    return false;
                }
                columns[j].add(board[i][j]); // add to the current column otherwise

                int box_index = (i / 3) * 3 + (j / 3); // we need to first calculate the indices of i, j to the box which is 3x3, since its integer it will automatically floor fractions
                if (boxes[box_index].contains(board[i][j])) { // if element already is in the box then its false
                    return false;
                }
                boxes[box_index].add(board[i][j]); // add to the box otherwise
            }
        }
        // if no duplicates found then its true
        return true;
    }
    public static void main(String[] args) {
        char[][] board = 
        {{'1','2','.','.','3','.','.','.','.'},
        {'4','.','.','5','.','.','.','.','.'},
        {'.','9','8','.','.','.','.','.','3'},
        {'5','.','.','.','6','.','.','.','4'},
        {'.','.','.','8','.','3','.','.','5'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','.','.','.','.','.','2','.','.'},
        {'.','.','.','4','1','9','.','.','8'},
        {'.','.','.','.','8','.','.','7','9'}}, 
        board2 = 
        {{'1','2','.','.','3','.','.','.','.'},
        {'4','.','.','5','.','.','.','.','.'},
        {'.','9','1','.','.','.','.','.','3'},
        {'5','.','.','.','6','.','.','.','4'},
        {'.','.','.','8','.','3','.','.','5'},
        {'7','.','.','.','2','.','.','.','6'},
        {'.','.','.','.','.','.','2','.','.'},
        {'.','.','.','4','1','9','.','.','8'},
        {'.','.','.','.','8','.','.','7','9'}};

        System.out.println(isValidSudoku(board));
        System.out.println(isValidSudoku(board2));

        // time and space complexity is O(1) because every sudoku board will be 9x9 meaning we always process 81 cells
    }
}

File: ArraysHashing/ValidSudoku.py
# Time Complexity: O(1)
# Space Complexity: O(1) 
'''
The time and space complexity will be O(1) since it'll always be a constant number of iterations, and the 
hashmaps will always have a constant size regardless of the input(as long as the input is bound by the constraints)
'''

def isValidSudoku(board):

    # O(1) since it'll always iterate 9 times regardless of the input
    # Iterate through the rows
    for row in board:
        rowMap = {}
        for cell in row:
            if(cell != '.' and cell in rowMap):
                return False
            else:
                rowMap[cell] = 0
    print("row")

    # O(1) since it'll always iterate 9 times regardless of the input
    # Iterate through the columns
    for col in range(len(board)):
        colMap = {}
        for row in range(len(board)):
            if(board[row][col] != '.' and board[row][col] in colMap):
                return False
            else:
                colMap[board[row][col]] = 0

    print("col")


    # O(1) since it'll always iterate 9 times regardless of the input
    # Iterate through the sub-boxes 
    rowLower, rowUpper = 0, 2
    colLower, colUpper = 0, 2
    for i in range(9):
        subBoxMap = {}
        for row in range(rowLower, rowUpper+1):
            for col in range(colLower, colUpper+1):
                print(subBoxMap, board[row][col])
                if(board[row][col] != '.' and board[row][col] in subBoxMap):
                    return False
                else:
                    subBoxMap[board[row][col]] = 0
        
        if(colUpper >= 8):
            rowLower += 3
            rowUpper += 3
            colLower, colUpper = 0, 2
        else:
            colLower += 3
            colUpper += 3
        


    
    print("sub-box")
    
    return True

print(isValidSudoku([["1","2",".",".","3",".",".",".","."],
 ["4",".",".","5",".",".",".",".","."],
 [".","9","8",".",".",".",".",".","3"],
 ["5",".",".",".","6",".",".",".","4"],
 [".",".",".","8",".","3",".",".","5"],
 ["7",".",".",".","2",".",".",".","6"],
 [".",".",".",".",".",".","2",".","."],
 [".",".",".","4","1","9",".",".","8"],
 [".",".",".",".","8",".",".","7","9"]]))

                












